sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousReceiveWork.java

```diff
@@ -97,95 +94,111 @@ int getNumberOfEvents() {
-            messageReceivedSink.next(message);
-            remaining.decrementAndGet();
-        } catch (Exception e) {
-            logger.warning("Exception occurred while publishing downstream.", e);
-            error(e);
+    synchronized boolean emitNext(ServiceBusReceivedMessage message) {
+        if (isTerminal.get()) {
+            return false;
         }
-    }
 
-    /**
-     * Completes the publisher. If the publisher has encountered an error, or an error has occurred, it does nothing.
-     */
-    void complete() {
-        logger.info("[{}]: Completing task.", id);
-        emitter.complete();
-        close();
-    }
+        if (!isStarted.get()) {
+            start();
+        }
 
-    /**
-     * Completes the publisher and sets the state to timeout.
-     */
-    void timeout() {
-        logger.info("[{}]: Work timeout occurred. Completing the work.", id);
-        emitter.complete();
-        workTimedOut = true;
-        close();
-    }
+        final int numberLeft = remaining.decrementAndGet();
 
-    /**
-     * Publishes an error downstream. This is a terminal step.
-     *
-     * @param error Error to publish downstream.
-     */
-    void error(Throwable error) {
-        this.error = error;
-        emitter.error(error);
-        close();
-    }
+        if (numberLeft < 0) {
+            logger.info("workId[{}] Number left {} < 0. Not emitting downstream.", id, numberLeft);
+            return false;
+        }
 
-    /**
-     * Returns the error object.
-     * @return the error.
-     */
-    Throwable getError() {
-        return this.error;
+        final Sinks.EmitResult result = downstreamEmitter.tryEmitNext(message);
+        if (result != Sinks.EmitResult.OK) {
+            logger.info("workId[{}] Could not emit downstream. EmitResult: {}", id, result);
+            return false;
+        }
+
+        // All events are emitted, so complete the synchronous work item. Next loop, it'll return false.
+        if (numberLeft == 0) {
+            complete(null);
+        }
+
+        return true;
     }
 
     /**
-     * Indiate that processing is started for this work.
+     * Completes the publisher.
+     *
+     * @param message Message to log.
      */
-    void startedProcessing() {
-        this.processingStarted = true;
+    synchronized void complete(String message) {
+        complete(message, null);
     }
 
     /**
+     * Completes the publisher. If the publisher has encountered an error, or an error has occurred, it does nothing.
      *
-     * @return flag indicting that processing is started or not.
+     * @param message Message to log. Null if there is no message to log.
+     * @param error Error if one occurred. Null otherwise.
      */
-    boolean isProcessingStarted() {
-        return this.processingStarted;
-    }
+    synchronized void complete(String message, Throwable error) {
+        if (isTerminal.getAndSet(true)) {
+            return;
+        }
```

Comment:
the `isTerminal` seems to ensure the `complete` body executes only once, if so do we need `complete` to be `synchronized`?

Reply:
also for the other `complete` overload.